### DISCUSSION TITLE
Extend RPI Example to Have 2 RPI's to One Ground Station

### DISCUSSION THREAD

**Discussion by Emil808** [nasa/fprime#1107] [Ideas]
I would like to extend the RPI example and have 2 RPIs connect to one ground station to act like a hub for both. I dug through the code and see that Telemetry, Events, and Commands are being streamed through the ByteStreamDriverModel component between the RPI and GDS with TCP being the communication protocol to send and receive. 

The TcpServer documentation says: 
> Designed for single client communication, it does not permit a queue of connecting clients.

So I would like some guidance on how and if the present TCP server implementation could be modified to queue connecting clients and service them one at a time. Would the TCP client component need to be modified also? 

Edit: 
Our company is working on an Nasa STTR Phase 2 for control and coordination of swarms of space vehicles (SV). Each SV will be identical in operation, and the GDS is to collect Events and telemetry, while and send commands either point to point, or broadcast. 

I made the diagram bellow to trace how Events, Telemetry, and Commands flow in and out of the RPI_Demo and to the ground station. 
For a first step, I would like to add a Device ID packer that will add the unique identifier before transmission. This will coincide with modifying the Byte stream Driver model implementation to handle multiple TCP clients on the server end, and then associating the client file descriptor to the Device ID at the deframer level on the GDS side. 

My current question is where can I modify the GDS to change the behavior of the TCP Server Component and Deframer? 

![BS_Comp_interaction](https://user-images.githubusercontent.com/54600430/145648495-7c1d17ef-7900-4c68-9562-9c6d2f5e556a.png)

**Comment by rawinza555** [comment]
Something you might want to keep in mind is that you might want to find a way to identify which telemetry is from which R PI. Maybe editing the framer and deframer to add an extra identifier for the Pi.

**Reply by Emil808** [reply]
Yes, I was thinking at which point does that identifier come into play. Thank you for suggesting the framer/deframer as a place to do that.

**Comment by Emil808** [comment]
In TcpSocketServer.hpp, the description for startup() on line 37 states: 

> Given the nature of this component, only one (1) client can be handled at a time. 

How exactly is the component only able to handle 1 client?

**Comment by Emil808** [comment]
TcpSocketServer inherited the IPSocket class which handles only one client connection by saving that file descriptor (fd) as a private member variable. To handle multiple clients in a single threaded tcp server, I could use fd_set, to handle multiple connections. I've been confused in which should I modify to accommodate fd_set. Which would be best to modify? the IPsocket class, or the TcpSocketServer class?

**Reply by LeStarch** [reply]
I would guess that the `TcpSocketServer` would be the place to edit as I don't think a `TcpClient` can handle multiple FDs.  I don't know a lot about the fd set, can it be used in place of a single FD seamlessly or does it need extra handling in the read/write calls?

IPSocket handles the read/writes from a FD, so you may need to update both.

**Reply by Emil808** [reply]
Whats confusing is that IPsocket holds one fd, but uses virtual functions implemented by its children. I'd want to change the behavior of just TcpSocketServer, so I shouldn't be changing anything within the IPsocket for TcpClient. So all the changes needed should be done within TcpSocketServer class. 

fd_select just indicates if any fd in the set is ready to be read from or written to. Then will need to use fd_isset to check if a specific fd is ready to be serviced. 

For a minimum of two clients (two fds), in context of sending an event, it will need to check with fd_isset if the receiving RPI is ready to receive, and if so, send the event (that is wrapped in the buffer when given into the TCP ServerComponent). If the receiving RPI is not ready according to fd_isset, server can go through n-attempts to keep checking fd_isset before indicating that the event couldn't be sent. the fd_set can be within the TCPSocketServer class, but how would that affect the fd in the IPsocket class? 

The next layer is deciding how to structure telemetry and events, to a specific RPI. Say, I have two RPI's running the same application but I want two different telemetry channels, and the ability to send Events to both or to just one. Things are handled by components. So would having a central component with a name like "Device_N" with N being a hard-programmed id number for each RPI work?

**Reply by LeStarch** [reply]
For tcp servers, there are 2 file descriptors that matter.  The one local to TcpServerSocket is the "server FD" used to work with the server connection (accept, listen).  Servers also have a FD for each client.  This is `m_fd` used in IPSocket as it is treated exactly like the FD on the client side code.

You primarily have two options:

1. Override all functions in IPSocket with a new class (e.g. MultiTcpServerSocket) and reimplement that code to handle the multiple connections effectively ignoring all of IPSocket except the interface
2. Override functions to IPSocket with something similar to the following read pseudo-code
```
if (fd_select()) {
   for (fd in client_fds) {
        if (fd_isset(fd)) {
            m_fd = fd;
            IpSocket::read()
        } 
   }
}
```

Option 2 takes advantage of the existing code, but it not the greatest form.  Option 1 would be cleaner code, but would involve starting over from scratch engineering multiple clients from the ground up.

**Reply by LeStarch** [reply]
Another thing to think about is sending: how do you know which client you are sending data too?  Are you going to track routing tokens?  Broadcast every command to every client?  On downlinked data, how are you going to track the source?  Reading multiple clients from one TcpServer will make the data less easy to understand.

Another approach that would keep these items clearly distinct would be to instantiate multiple TcpServer components, each on its own port, and have one client per-server.

**Reply by Emil808** [reply]
Thanks for the guidance!
For the approach you suggested, how would that scale when wanting to connect to more than 10 devices? At which file can I implement that?
